pkg/services/govmomi/power.go

```diff
@@ -0,0 +1,108 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package govmomi
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha1"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"                         //nolint
+	infrav1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/v1beta1" //nolint
+	"sigs.k8s.io/cluster-api/util/conditions"                       //nolint
+)
+
+func (vms *VMService) getPowerState(ctx *virtualMachineContext) (infrav1.VirtualMachinePowerState, error) {
+	powerState, err := ctx.Obj.PowerState(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	switch powerState {
+	case types.VirtualMachinePowerStatePoweredOn:
+		return infrav1.VirtualMachinePowerStatePoweredOn, nil
+	case types.VirtualMachinePowerStatePoweredOff:
+		return infrav1.VirtualMachinePowerStatePoweredOff, nil
+	case types.VirtualMachinePowerStateSuspended:
+		return infrav1.VirtualMachinePowerStateSuspended, nil
+	default:
+		return "", errors.Errorf("unexpected power state %q for vm %s", powerState, ctx)
+	}
+}
+
+func (vms *VMService) isSoftPowerOffTimeoutExceeded(vm *infrav1.VSphereVM) bool {
+	if !conditions.Has(vm, infrav1.GuestSoftPowerOffCondition) {
+		// The SoftPowerOff never got triggered, so it can't be timed out yet.
+		return false
+	}
+	if vm.Spec.PowerOffMode == vmopv1.VirtualMachinePowerOpModeSoft {
+		// Timeout only applies to trySoft mode.
+		// For soft mode it will wait infinitely.
+		return false
+	}
+	now := time.Now()
+	timeSoftPowerOff := conditions.GetLastTransitionTime(vm, infrav1.GuestSoftPowerOffCondition)
+	diff := now.Sub(timeSoftPowerOff.Time)
+	var timeout time.Duration
+	if vm.Spec.GuestSoftPowerOffTimeout != nil {
+		timeout = vm.Spec.GuestSoftPowerOffTimeout.Duration
+	} else {
+		timeout = infrav1.GuestSoftPowerOffDefaultTimeout
+	}
+	return timeout.Seconds() > 0 && diff.Seconds() >= timeout.Seconds()
+}
+
+// triggerSoftPowerOff tries to trigger a soft power off for a VM to shut down the guest.
+// It returns true if the soft power off operation is pending.
+func (vms *VMService) triggerSoftPowerOff(ctx *virtualMachineContext) (bool, error) {
+	if ctx.VSphereVM.Spec.PowerOffMode == vmopv1.VirtualMachinePowerOpModeHard {
+		// hard power off is expected.
+		return false, nil
+	}
+
+	if conditions.Has(ctx.VSphereVM, infrav1.GuestSoftPowerOffCondition) {
+		// soft power off operation has been triggered.
+		if ctx.VSphereVM.Spec.PowerOffMode == vmopv1.VirtualMachinePowerOpModeSoft {
+			return true, nil
+		}
+
+		return !vms.isSoftPowerOffTimeoutExceeded(ctx.VSphereVM), nil
+	}
+
+	vmwareToolsRunning, err := ctx.Obj.IsToolsRunning(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	if !vmwareToolsRunning {
+		// VMware tools is not installed. Returning false to force a power off.
+		return false, nil
```

Comment:
We should not `return false, nil` when the mode is set to `Soft` if I get the definition of `Soft` right.

Otherwise `Soft` is `TrySoft` but with an unlimited timeout only. In this case I'm thinking about if we should even add a `Soft` mode (open for opinions).

Or do we have to re-define Soft, that it will do a hard shutoff, if VMwareTools are not running or `GUsetStateChangeSupported` is not `true`?

Reply:
Note: I'm okay with both, but we have to define it clearly in the godoc then :-)

Reply:
Checked the VM Operator and it will not try to power off the VM if it is set to `soft`.
Will keep the same behavior.